.design/requirements-575.md

```diff
@@ -0,0 +1,269 @@
+
+    @OnError
+    public void onError(
+        @PathParam("targetId") String targetId,
+        Session session,
+        Throwable error) {
+        // Error handling the WebSocket stream
+    }
+
+    @OnMessage
+    public void onMessage(
+        @PathParam("targetId") String targetId,
+        String message,
+        Session session) {
+        // Process incoming message
+		// The "message" argument can also be of type Reader, Decoder.Text
+		// or Decoder.TextStream.
+		// See @OnMessage documentation for more details.
+    }
+
+	@OnMessage
+    public void onMessage(
+        @PathParam("targetId") String targetId,
+        byte[] message,
+        Session session) {
+        // Process binary message
+		// The "message" argument can also be of type ByteBuffer, InputStream,
+		// Decoder.Binary or Decoder.BinaryStream.
+		// See @OnMessage documentation for more details.
+    }
+
+    @OnMessage
+    public void onPong(
+        @PathParam("targetId") String targetId,
+        PongMessage message,
+        Session session) {
+        // Process pong message
+    }
+}
+```
+
+#### WebSocket pattern description
+
+In a Jakarta environment, we would annotate this class with `@ServerEndpoint("/ws/work/{targetId}")` and register it in a Jakarta WebSocket-ready HTTP server.
+
+In OSGi, we might want to use annotations similar to the ones in the Jarkata Servlet whiteboard specification to provide more control, *e.g.* `@WebSocketWhiteboardServletPattern("/ws/work/{targetId}")`.
+
+#### WebSocket subprotocols
+
+The [WebSocket specification (RFC 6455)](https://datatracker.ietf.org/doc/html/rfc6455#section-1.9) indicates that clients can request, during the handshake, can give a list of subprotocols names they expect, in order of preference.
+If a list of subprotocols is specified, the server must use and return the subprotocol it selected from that list.
+
+RFC 6455 doesn't explicitly specifies what happens if the server doesn't support any of the client subprotocols.
+
+The list of subprotocols supported by the OSGi WebSocket Whiteboard server endpoint can be given using a service property of type `string[]`.
+A class annotation, *e.g.* `WebSocketWhiteboardSubprotocols`, could be used to ease the declaration of such list.
+
+#### Service scope
+
+The Jakarta WebSocket specification expects one instance of WebSocket handler per session.
+As a result, it is expected by the OSGi WebSocket Whiteboard specification to consume services with the `PROTOTYPE` service scope, to obtain a new instance of the service per session.
+
+Implementations using the `SINGLETON` service scope must manage thread safety and session handling by themselves.
+The `BUNDLE` service scope should be considered as a `SINGLETON`.
+
+### UC2: Define and register a WebSocket server endpoint using the `Endpoint` class
+
+This is the equivalent of [UC1](#uc1-define-and-register-a-websocket-server-endpoint-using-annotations) but extending a class instead of using annotations.
```

Comment:
Same as above, there must only one instance per server endpoint! In the sense of the Whiteboard pattern I won't support programatic endpoints at al as annotation based ones already supply what needed here.